aptos-move/move-examples/token_objects/ambassador/move/sources/ambassador.move

```diff
@@ -0,0 +1,277 @@
+        object::transfer_with_ref(linear_transfer_ref, soul_bound_to);
+
+        // Disables ungated transfer, thus making the token soulbound
+        object::disable_ungated_transfer(&transfer_ref);
+
+        // Initialize the ambassador token
+        let ambassador_token = AmbassadorToken {
+            burn_ref,
+            property_mutator_ref,
+            level_update_events: object::new_event_handle(&object_signer),
+        };
+        move_to(&object_signer, ambassador_token);
+
+        // Initializes the ambassador level
+        move_to(&object_signer, AmbassadorLevel { ambassador_level: 0 });
+
+        // Initialize the property map with the rank property set to 1
+        let property_keys = vector<String>[string::utf8(b"Rank")];
+        let property_types = vector<String>[string::utf8(b"0x1::string::String")];
+        let property_values = vector<vector<u8>>[bcs::to_bytes<String>(&string::utf8(RANK_BRONZE))];
+        let properties = property_map::prepare_input(
+            property_keys,
+            property_types,
+            property_values,
+        );
+        property_map::init(&constructor_ref, properties);
+    }
+
+    /// Burns an ambassador token
+    public entry fun burn(creator: &signer, token: Object<AmbassadorToken>) acquires AmbassadorToken {
+        authorize_creator(creator, &token);
+        let ambassador_token = move_from<AmbassadorToken>(object::object_address(&token));
+        let AmbassadorToken {
+            burn_ref,
+            property_mutator_ref,
+            level_update_events,
+        } = ambassador_token;
+        event::destroy_handle(level_update_events);
+        property_map::burn(property_mutator_ref);
+        token::burn(burn_ref);
+    }
+
+    #[view]
+    /// Returns the ambassador level of the token
+    public fun ambassador_level(token: Object<AmbassadorToken>): u64 acquires AmbassadorLevel {
+        let ambassador_level = borrow_global<AmbassadorLevel>(object::object_address(&token));
+        ambassador_level.ambassador_level
+    }
+
+    /// Sets the ambassador level of the token
+    public entry fun set_ambassador_level(
+        creator: &signer,
+        token: Object<AmbassadorToken>,
+        new_ambassador_level: u64
+    ) acquires AmbassadorLevel, AmbassadorToken {
+        authorize_creator(creator, &token);
+        let token_address = object::object_address(&token);
+        let ambassador_level = borrow_global_mut<AmbassadorLevel>(token_address);
+        event::emit_event(
+            &mut borrow_global_mut<AmbassadorToken>(token_address).level_update_events,
+            LevelUpdateEvent {
+                token_address,
+                old_level: ambassador_level.ambassador_level,
+                new_level: new_ambassador_level,
+            }
+        );
+        ambassador_level.ambassador_level = new_ambassador_level;
+        update_ambassador_rank(token, new_ambassador_level);
+    }
+
+    #[view]
```

Comment:
nit: Let's keep a consistent order of functions in this file:
1. Constructor: init_module
2. View/getters: view functions
3. Create/destroy: mint/burn tokens
4. Setters: set level/set rank

Reply:
Done.